olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -286,5 +305,158 @@ CancellationToken OlpClient::CallApi(
       [cancel_context]() { cancel_context->CancelOperation(); });
 }
 
+std::tuple<http::NetworkResponse, std::stringstream> SendRequest(
+    const http::NetworkRequest& request,
+    const olp::client::OlpClientSettings& settings,
+    const olp::client::RetrySettings& retry_settings,
+    client::CancellationContext context) {
+  http::NetworkResponse network_response = kCancelledErrorResponse;
+  auto interest_flag = std::make_shared<std::atomic_bool>(true);
+  Condition condition{};
+  auto network = settings.network_request_handler;
+  auto response_body = std::make_shared<std::stringstream>();
+
+  http::SendOutcome outcome =
+      network->Send(request, response_body,
+                    [&, interest_flag](http::NetworkResponse response) {
+                      if (interest_flag->exchange(false)) {
+                        network_response = std::move(response);
+                        condition.Notify();
+                      }
+                    });
+
+  if (outcome.IsSuccessful()) {
+    context.ExecuteOrCancelled(
+        [&, interest_flag]() {
```

Comment:
Not sure I get this. The first lambda is executed sync, only the second lambda passed to CancellationToken might be triggered async by the user.

Different idea. Maybe worth it wrapping the network->Send() into the first lambda?

```cpp
context.ExecuteOrCancelled( [&]() {
    http::SendOutcome outcome = network->Send(request, response_body,
                    [&, interest_flag](http::NetworkResponse response) {
                      if (interest_flag->exchange(false)) {
                        network_response = std::move(response);
                        condition.Notify();
                      }
                    });
 
   ......   
 
    return CancellationToken([&, interest_flag]() {
            if (interest_flag->exchange(false)) {
              network->Cancel(outcome.GetRequestId());
              network_response = kCancelledErrorResponse;
              network_response.WithRequestId(outcome.GetRequestId());
              condition.Notify();
            }
          });
```

Reply:
Yes, my bad. We may pass interest_flag in ExecuteOrCancelled lambda by reference.

Reply:
And yes, it would be good to wrap Send in ExecuteOrCancelled, but it should be separate clousre.